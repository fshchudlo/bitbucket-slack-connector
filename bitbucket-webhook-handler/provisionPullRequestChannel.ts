import { SlackAPIAdapter } from "./ports/SlackAPIAdapter";
import { BitbucketNotification, PullRequestBasicNotification } from "../bitbucket-payload-types";
import { WebhookHandlerConfig } from "./webhookHandlerConfig";
import { buildChannelName } from "./buildChannelName";
import handleBitbucketWebhook from "./handleBitbucketWebhook";

export async function provisionPullRequestChannel(slackAPI: SlackAPIAdapter, payload: BitbucketNotification, config: WebhookHandlerConfig) {
    const channelName = buildChannelName(payload.pullRequest);
    if (payload.eventKey == "pr:opened") {
        return await slackAPI.createChannel({ name: channelName, isPrivate: config.usePrivateChannels });
    }
    const existingChannel = await slackAPI.findChannel(channelName, config.usePrivateChannels);
    if (existingChannel != null) {
        return existingChannel;
    }

    const prOpenedPayload = <PullRequestBasicNotification>{
        eventKey: "pr:opened",
        actor: {
            displayName: payload.pullRequest.author.user.displayName,
            emailAddress: payload.pullRequest.author.user.emailAddress
        },
        pullRequest: payload.pullRequest
    };
    await handleBitbucketWebhook(prOpenedPayload, slackAPI, config);
    return await slackAPI.findChannel(channelName, config.usePrivateChannels);
}