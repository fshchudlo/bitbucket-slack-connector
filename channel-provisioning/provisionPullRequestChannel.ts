import { SlackNotificationChannel } from "../bitbucket-webhook-handler/SlackNotificationChannel";
import { BitbucketNotification, PullRequestBasicNotification } from "../bitbucket-payload-types";
import { WebhookHandlerConfig } from "../bitbucket-webhook-handler/webhookHandlerConfig";
import { buildChannelName } from "./buildChannelName";
import handleBitbucketWebhook from "../bitbucket-webhook-handler/handleBitbucketWebhook";
import { SlackChannelFactory } from "./SlackChannelFactory";

export async function provisionPullRequestChannel(channelFactory: SlackChannelFactory, slackAPI: SlackNotificationChannel, payload: BitbucketNotification, config: WebhookHandlerConfig) {
    const channelName = buildChannelName(payload.pullRequest);
    if (payload.eventKey == "pr:opened") {
        return await channelFactory.createChannel({ name: channelName, isPrivate: config.usePrivateChannels });
    }
    const existingChannel = await channelFactory.findChannel(channelName, config.usePrivateChannels);
    if (existingChannel != null) {
        return existingChannel;
    }

    const newChannel = await channelFactory.createChannel({ name: channelName, isPrivate: config.usePrivateChannels });
    const prOpenedPayload = <PullRequestBasicNotification>{
        eventKey: "pr:opened",
        actor: {
            displayName: payload.pullRequest.author.user.displayName,
            emailAddress: payload.pullRequest.author.user.emailAddress
        },
        pullRequest: payload.pullRequest
    };
    await handleBitbucketWebhook(prOpenedPayload, slackAPI, newChannel, config);
    return await channelFactory.findChannel(channelName, config.usePrivateChannels);
}