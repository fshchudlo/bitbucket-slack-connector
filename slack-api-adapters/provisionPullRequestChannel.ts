import { BitbucketNotification, PullRequestBasicNotification } from "../bitbucket-payload-types";
import { buildChannelName } from "./buildChannelName";
import { SlackChannelFactory } from "./slack-channel-factory/SlackChannelFactory";
import handleWebhookPayload from "../bitbucket-webhook-handler/handleWebhookPayload";
import { SlackChannel } from "../bitbucket-webhook-handler/SlackChannel";
import { AppConfig } from "../app.config";

export async function provisionPullRequestChannel(channelFactory: SlackChannelFactory, broadcastChannel: SlackChannel | null, payload: BitbucketNotification) {
    const channelName = buildChannelName(payload.pullRequest);
    if (payload.eventKey == "pr:opened") {
        return await channelFactory.setupNewChannel({
            name: channelName,
            isPrivate: AppConfig.USE_PRIVATE_CHANNELS,
            defaultParticipants: AppConfig.DEFAULT_CHANNEL_PARTICIPANTS
        });
    }
    const existingChannel = await channelFactory.fromExistingChannel(channelName, AppConfig.USE_PRIVATE_CHANNELS);
    if (existingChannel != null) {
        return existingChannel;
    }

    const createdChannel = await channelFactory.setupNewChannel({
        name: channelName,
        isPrivate: AppConfig.USE_PRIVATE_CHANNELS,
        defaultParticipants: AppConfig.DEFAULT_CHANNEL_PARTICIPANTS
    });
    const prOpenedPayload = <PullRequestBasicNotification>{
        eventKey: "pr:opened",
        actor: {
            displayName: payload.pullRequest.author.user.displayName,
            emailAddress: payload.pullRequest.author.user.emailAddress
        },
        pullRequest: payload.pullRequest
    };
    await handleWebhookPayload(prOpenedPayload, createdChannel, broadcastChannel);
    return createdChannel;
}