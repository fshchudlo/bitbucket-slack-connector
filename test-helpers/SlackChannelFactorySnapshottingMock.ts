import { SlackChannel } from "../bitbucket-webhook-handler/SlackChannel";
import {
    CreateChannelArguments,
    SlackChannelFactory,
    SlackChannelInfo
} from "../channel-provisioning/SlackChannelFactory";
import SlackChannelSnapshottingMock from "./SlackChannelSnapshottingMock";

const channelId = "12345";
export default class SlackChannelFactorySnapshottingMock implements SlackChannelFactory {
    snapshot: {
        createdChannels: SlackChannelInfo[];
        searchedChannels: any[];
    };

    constructor() {
        this.snapshot = {
            createdChannels: new Array<SlackChannelInfo>(),
            searchedChannels: new Array<any>()
        };
    }

    fromExistingChannel(channelName: string, includePrivateChannels: boolean): Promise<SlackChannel | null> {
        this.snapshot.searchedChannels.push({ channelName, includePrivateChannels });
        const channelInfo = this.snapshot.createdChannels.find(c => c.name == channelName);
        return Promise.resolve(channelInfo ? new SlackChannelSnapshottingMock() : null);
    }

    setupNewChannel(options: CreateChannelArguments): Promise<SlackChannel> {
        const channelInfo = <SlackChannelInfo>{ id: channelId, name: options.name, isArchived: false };
        this.snapshot.createdChannels.push(channelInfo);
        return Promise.resolve(new SlackChannelSnapshottingMock());
    }
}
