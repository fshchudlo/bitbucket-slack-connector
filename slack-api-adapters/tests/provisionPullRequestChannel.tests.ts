import { provisionPullRequestChannel } from "../provisionPullRequestChannel";
import TestPayloadBuilder from "../../test-helpers/TestPayloadBuilder";
import SlackChannelFactorySnapshottingMock from "./SlackChannelFactorySnapshottingMock";

describe("provisionPullRequestChannel", () => {
    it("Should create channel", async () => {
        const testFactory = new SlackChannelFactorySnapshottingMock();

        await provisionPullRequestChannel(testFactory, null, TestPayloadBuilder.pullRequestOpened());

        expect(testFactory.snapshot).toMatchSnapshot();
    });

    it("Should replay channel creation if channel doesn't exist and handle initial payload after that", async () => {
        const testFactory = new SlackChannelFactorySnapshottingMock();

        await provisionPullRequestChannel(testFactory, null, TestPayloadBuilder.reviewersUpdated());

        expect(testFactory.snapshot).toMatchSnapshot();
    });
});
