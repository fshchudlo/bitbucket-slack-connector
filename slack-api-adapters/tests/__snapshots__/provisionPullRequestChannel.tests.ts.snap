// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`provisionPullRequestChannel Should create channel 1`] = `
{
  "createdChannels": [
    {
      "id": "12345",
      "isArchived": false,
      "name": "pr-test-proj-test-repository-1",
    },
  ],
  "searchedChannels": [],
}
`;

exports[`provisionPullRequestChannel Should replay channel creation if channel doesn't exist and handle initial payload after that 1`] = `
{
  "createdChannels": [
    {
      "id": "12345",
      "isArchived": false,
      "name": "pr-test-proj-test-repository-1",
    },
  ],
  "searchedChannels": [
    {
      "channelName": "pr-test-proj-test-repository-1",
      "includePrivateChannels": true,
    },
  ],
}
`;
